src/java.desktop/macosx/classes/apple/laf/JRSUIControl.java

```diff
@@ -109,16 +113,29 @@ public JRSUIControl(final boolean flipped){
         flipped = other.flipped;
         cfDictionaryPtr = getCFDictionary(flipped);
         if (cfDictionaryPtr == 0) throw new RuntimeException("Unable to create native representation");
+        Disposer.addRecord(disposerReferent, new JRSUIControlDisposerRecord(cfDictionaryPtr));
         nativeMap = new HashMap<Key, DoubleValue>();
         changes = new HashMap<Key, DoubleValue>(other.nativeMap);
         changes.putAll(other.changes);
     }
 
-    @SuppressWarnings("removal")
-    protected synchronized void finalize() throws Throwable {
-        if (cfDictionaryPtr == 0) return;
-        disposeCFDictionary(cfDictionaryPtr);
-        cfDictionaryPtr = 0;
+    private static class JRSUIControlDisposerRecord implements DisposerRecord {
+
+        private long cfDictionaryPtr;
+        JRSUIControlDisposerRecord(long ptr) {
+            cfDictionaryPtr = ptr;
+        }
+
+        public synchronized void dispose() {
+            if (cfDictionaryPtr == 0) {
+                return;
+            }
+            try {
+                disposeCFDictionary(cfDictionaryPtr);
+            } catch (Throwable t) {
+            }
+            cfDictionaryPtr = 0;
+        }
```

Comment:
Is it possible for this method to be called twice? It seems that not all other implementations include the ptr != 0 guard.

Also, synchronization appears to be used on the wrong object. I believe it should be synchronized on the outer JRSUIControl class, which uses synchronization to access the cfDictionaryPtr field. Alternatively, we might consider implementing synchronization in a different way.

Reply:
That would be great if we could drop all unneeded code.